In [ ]:
# YOLOv11 Training Script with Full Parameter Explanations and Improved Stability

from ultralytics import YOLO
import os
from datetime import datetime

# --- Model Setup ---
# Load a pretrained YOLOv8n model (nano version: fastest, smallest, good for limited data or edge devices)
model = YOLO('../src/human_detector/models/yolov11s.pt')
# To resume training from a previous checkpoint:
# model = YOLO('/path/to/your/best.pt')

# --- Training Configuration ---
epochs = 300               # Number of times the model sees the entire dataset
imgsz = 640                # Size to which all images are resized (square). 640 is a good trade-off between speed and accuracy
batch_size = 4             # Number of images used in one forward/backward pass (small for limited VRAM)

lr0 = 1e-5                 # Initial learning rate (controls how much the model updates each step)
lrf = 0.01                 # Final learning rate = lr0 * lrf, used with cosine decay learning schedule
optimizer = "Adam"          # Optimizer algorithm: 'SGD' is more stable for small datasets, 'AdamW' (default) adapts better to large, noisy datasets
# warmup_epochs = 3          # Number of epochs to slowly ramp up learning rate to avoid instability at start
dropout = 0.1              # Dropout rate (randomly turns off neurons during training to prevent overfitting)

# --- Experiment Folder Naming ---
time_stamp = datetime.now().strftime("%Y%m%d_%H%M")

exp_name = (
    f"yolov11s_v1_"         # Model name and version. You can use this to identify custom model variants.
    f"ep{epochs}_"          # Number of epochs (e.g., ep300 means trained for 300 full dataset passes)
    f"img{imgsz}_"          # Image input size (e.g., img640 means all images resized to 640x640)
    f"bs{batch_size}_"      # Batch size (e.g., bs4 means 4 images processed per training step)
    f"lr{lr0}_"             # Initial learning rate (e.g., lr0.0001)
    f"lrf{lrf}_"            # Final learning rate factor (e.g., lrf0.01 → lr drops to 0.000001 at end)
    f"{optimizer}_"         # Optimizer used (e.g., SGD, AdamW, etc.)
    f"augmented_"            # Tag for augmentation setup (e.g., Final or custom profile name)
    f"{time_stamp}"         # Timestamp for uniqueness and chronological sorting
)

# --- Training Start ---
results = model.train(
    seed=42,                     # Random seed for reproducibility
    data='data.yaml',            # Path to data config file (defines class names and image paths)
    epochs=epochs,               # Total number of training epochs
    patience=10,                 # Early stopping: if val loss doesn't improve for 10 epochs, stop training
    dropout=dropout,             # Helps prevent overfitting by randomly disabling neurons
    imgsz=imgsz,                 # Resizes all images to this square dimension
    batch=batch_size,            # Number of images per training batch
    lr0=lr0,                     # Starting learning rate
    lrf=lrf,                     # Final learning rate multiplier
    # warmup_epochs=warmup_epochs,# Smooth ramp-up of LR to avoid unstable gradients
    optimizer=optimizer,         # Optimizer choice: SGD recommended for small datasets
    device=0,                    # GPU device ID (0 = first GPU)
    half=True,                   # Use float16 precision if supported (faster training)
    workers=2,                   # Number of CPU workers for data loading (adjust based on CPU)
    cache=True,                  # Preload and cache images in memory for faster epoch execution
    rect=True,                   # Enables rectangular training for better image aspect ratio handling
    verbose=True,                # Print detailed training logs for every batch
    amp=False ,
    # exist_ok=True,               # Overwrites existing output folder if it already exists

    # --- Data Augmentation (tuned for small datasets) ---
    hsv_h=0.015,                 # Hue variation (color tint)
    hsv_s=0.7,                   # Saturation variation (color strength)
    hsv_v=0.4,                   # Brightness variation
    translate=0.05,              # Random image shift (reduced for safety)
    scale=0.1,                   # Random zoom in/out (less aggressive)
    fliplr=0.5,                  # 50% chance to flip images horizontally
    mosaic=0.2,                  # Combines 4 training images into one
    mixup=0.03,                  # Blends two images and labels
    copy_paste=0.03,             # Copies object from one image and pastes onto another
    perspective=0.001,           # Small perspective distortion to improve robustness

    # --- Output Settings ---
    project='results',           # Base directory to store training logs and weights
    name=exp_name                # Subfolder name under project/ to store this run’s output
)

# --- Save Path Display ---
save_dir = str(results.save_dir)
best_pt = os.path.join(save_dir, 'weights', 'best.pt')
print(f"✅ Trained checkpoint saved to: {best_pt}")


New https://pypi.org/project/ultralytics/8.3.146 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.141 🚀 Python-3.10.12 torch-2.7.0+cu126 CUDA:0 (NVIDIA GeForce MX550, 1864MiB)
engine/trainer: agnostic_nms=False, amp=False, augment=False, auto_augment=randaugment, batch=4, bgr=0.0, box=7.5, cache=True, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.03, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=data.yaml, degrees=0.0, deterministic=True, device=0, dfl=1.5, dnn=False, dropout=0.1, dynamic=False, embed=None, epochs=300, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=True, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=1e-05, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.03, mode=train, model=../src/human_detector/models/yolov11n.pt, momentum=0.937, mosaic=0.2, multi_scale=False, name=yolov11n_v1_ep300_img640_bs4_l

train: Scanning /home/yogee/Desktop/human_detector_ws/data_training/data_person/train/labels.cache... 4044 images, 23 backgrounds, 0 corrupt: 100%|██████████| 4044/4044 [00:00<?, ?it/s]


WARNING ⚠️ cache='ram' may produce non-deterministic training results. Consider cache='disk' as a deterministic alternative if your disk space allows.


train: Caching images (3.3GB RAM): 100%|██████████| 4044/4044 [00:12<00:00, 320.04it/s] 

WARNING ⚠️ 'rect=True' is incompatible with DataLoader shuffle, setting shuffle=False


val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 317.9±125.2 MB/s, size: 448.1 KB)


val: Scanning /home/yogee/Desktop/human_detector_ws/data_training/data_person/test/labels.cache... 1720 images, 44 backgrounds, 0 corrupt: 100%|██████████| 1720/1720 [00:00<?, ?it/s]


WARNING ⚠️ cache='ram' may produce non-deterministic training results. Consider cache='disk' as a deterministic alternative if your disk space allows.


val: Caching images (1.4GB RAM): 100%|██████████| 1720/1720 [00:07<00:00, 239.86it/s]


Plotting labels to results/yolov11n_v1_ep300_img640_bs4_lr1e-05_lrf0.01_Adam_augmented_20250528_2232/labels.jpg... 
optimizer: Adam(lr=1e-05, momentum=0.937) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to results/yolov11n_v1_ep300_img640_bs4_lr1e-05_lrf0.01_Adam_augmented_20250528_2232
Starting training for 300 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      1/300     0.715G       1.42      1.794       1.56         19        384: 100%|██████████| 1011/1011 [02:45<00:00,  6.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:22<00:00,  9.73it/s]


                   all       1720       3999      0.405      0.543      0.452      0.234

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      2/300      1.67G      1.356      1.809      1.509         18        384: 100%|██████████| 1011/1011 [02:50<00:00,  5.93it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.97it/s]


                   all       1720       3999       0.72      0.544      0.622      0.388

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      3/300      1.58G      1.277      1.756      1.463         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.95it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.90it/s]


                   all       1720       3999      0.627      0.616      0.699      0.495

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      4/300      1.58G      1.172      1.509      1.384         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.95it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.00it/s]


                   all       1720       3999      0.781      0.606      0.725      0.513

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      5/300      1.58G       1.12      1.418      1.347         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.97it/s]


                   all       1720       3999      0.803       0.63      0.747      0.532

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      6/300      1.58G      1.101      1.356       1.33         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.13it/s]


                   all       1720       3999      0.835      0.633       0.76      0.529

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      7/300      1.66G      1.076      1.309      1.309         18        384: 100%|██████████| 1011/1011 [02:48<00:00,  5.98it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.02it/s]


                   all       1720       3999      0.837      0.653       0.77      0.543

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      8/300      1.57G      1.053      1.269      1.292         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.04it/s]


                   all       1720       3999      0.845      0.663      0.778      0.547

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      9/300      1.66G      1.043      1.241      1.283         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  5.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.09it/s]


                   all       1720       3999      0.847      0.669      0.787      0.551

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     10/300      1.58G      1.027      1.212      1.274         19        384: 100%|██████████| 1011/1011 [03:13<00:00,  5.23it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.92it/s]


                   all       1720       3999      0.853      0.679      0.795      0.554

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     11/300      1.58G      1.018      1.191      1.263         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.97it/s]


                   all       1720       3999      0.857      0.683      0.802      0.572

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     12/300      1.58G      1.003      1.163      1.254         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.98it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.94it/s]


                   all       1720       3999       0.85        0.7       0.81      0.561

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     13/300      1.66G     0.9917      1.143      1.244         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  5.98it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.91it/s]


                   all       1720       3999      0.854      0.697      0.816      0.574

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     14/300      1.66G     0.9794      1.116      1.239         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  5.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.03it/s]


                   all       1720       3999      0.847      0.708       0.82      0.561

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     15/300      1.66G     0.9674      1.099      1.229         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.90it/s]


                   all       1720       3999      0.847      0.722      0.823      0.587

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     16/300      1.57G     0.9567      1.082      1.229         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.93it/s]


                   all       1720       3999      0.842      0.743       0.83      0.591

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     17/300      1.66G     0.9593       1.07       1.22         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.93it/s]


                   all       1720       3999      0.869      0.739      0.835       0.58

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     18/300      1.66G     0.9461      1.049       1.21         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.98it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.94it/s]


                   all       1720       3999       0.86      0.753      0.839      0.579

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     19/300      1.57G     0.9374      1.038      1.207         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.97it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00,  9.94it/s]


                   all       1720       3999      0.858      0.761      0.841      0.599

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     20/300      1.58G      0.936      1.021      1.209         19        384: 100%|██████████| 1011/1011 [02:49<00:00,  5.96it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.09it/s]


                   all       1720       3999      0.883      0.766      0.849       0.59

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     21/300      1.58G     0.9364      1.013      1.208         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.14it/s]


                   all       1720       3999      0.868      0.775      0.851      0.582

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     22/300      1.58G      0.915     0.9978      1.195         19        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.09it/s]


                   all       1720       3999      0.877      0.783      0.856      0.597

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     23/300      1.66G     0.9211     0.9893      1.197         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.11it/s]


                   all       1720       3999      0.881      0.785      0.858        0.6

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     24/300      1.57G     0.9071      0.974      1.189         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.09it/s]


                   all       1720       3999      0.875       0.79       0.86      0.613

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     25/300      1.66G     0.9099     0.9564      1.191         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.10it/s]


                   all       1720       3999      0.883      0.796      0.864      0.619

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     26/300      1.66G     0.9035     0.9491      1.185         19        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.15it/s]


                   all       1720       3999       0.88      0.801      0.866      0.595

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     27/300      1.57G     0.8974     0.9372      1.179         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:20<00:00, 10.25it/s]


                   all       1720       3999       0.87      0.806      0.863      0.598

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     28/300      1.66G      0.895     0.9263      1.176         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.20it/s]


                   all       1720       3999      0.865      0.813      0.868      0.593

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     29/300      1.57G     0.8896     0.9177      1.171         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.14it/s]


                   all       1720       3999      0.877      0.813      0.872      0.597

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     30/300      1.58G     0.8945     0.9124      1.174         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.15it/s]


                   all       1720       3999      0.881      0.816      0.873      0.626

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     31/300      1.58G     0.8773      0.898       1.16         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.22it/s]


                   all       1720       3999      0.892      0.811      0.874      0.626

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     32/300      1.58G     0.8742     0.8903      1.164         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.15it/s]


                   all       1720       3999      0.892      0.814      0.874      0.614

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     33/300      1.66G     0.8697      0.884      1.154         19        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.13it/s]


                   all       1720       3999      0.894      0.815      0.876      0.613

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     34/300      1.57G     0.8703     0.8716      1.158         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.17it/s]


                   all       1720       3999      0.889      0.819      0.877      0.604

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     35/300      1.66G     0.8646     0.8658      1.152         19        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.20it/s]


                   all       1720       3999      0.895      0.818      0.878        0.6

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     36/300      1.58G     0.8736     0.8624      1.153         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.18it/s]


                   all       1720       3999      0.896      0.815      0.878      0.635

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     37/300      1.66G     0.8512     0.8536      1.146         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.16it/s]


                   all       1720       3999      0.895      0.824       0.88      0.641

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     38/300      1.57G     0.8429     0.8391      1.144         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.17it/s]


                   all       1720       3999      0.897      0.822      0.883      0.641

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     39/300      1.66G     0.8409     0.8347      1.141         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  5.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.18it/s]


                   all       1720       3999      0.899      0.825      0.883      0.629

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     40/300      1.57G     0.8315     0.8304      1.136         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  5.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.14it/s]


                   all       1720       3999      0.895      0.826      0.883      0.634

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     41/300      1.66G      0.834     0.8244      1.133         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.21it/s]


                   all       1720       3999      0.896       0.83      0.886      0.641

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     42/300      1.66G     0.8346     0.8144      1.135         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  5.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.14it/s]


                   all       1720       3999      0.894      0.826      0.882      0.629

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     43/300      1.57G     0.8199     0.8066      1.126         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.13it/s]


                   all       1720       3999      0.901      0.825      0.886      0.618

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     44/300      1.66G     0.8367     0.8073      1.133         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:20<00:00, 10.24it/s]


                   all       1720       3999      0.896       0.83      0.885      0.638

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     45/300      1.58G     0.8275      0.801      1.125         17        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.12it/s]


                   all       1720       3999      0.896      0.836      0.889      0.649

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     46/300      1.58G     0.8236     0.7912      1.125         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.15it/s]


                   all       1720       3999        0.9      0.834      0.889      0.653

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     47/300      1.58G     0.8112     0.7824       1.12         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.17it/s]


                   all       1720       3999      0.901      0.837      0.889      0.642

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     48/300      1.58G     0.8263     0.7822      1.121         18        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.16it/s]


                   all       1720       3999      0.901      0.828      0.887      0.645

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     49/300      1.66G     0.8205     0.7732      1.122         18        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.16it/s]


                   all       1720       3999      0.904      0.834      0.892       0.64

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     50/300      1.57G     0.8156     0.7715      1.118         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  5.99it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.12it/s]


                   all       1720       3999      0.903      0.833       0.89      0.638

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     51/300      1.66G     0.8061     0.7648      1.115         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.16it/s]


                   all       1720       3999      0.897       0.84       0.89      0.643

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     52/300      1.58G     0.8177     0.7672      1.114         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.19it/s]


                   all       1720       3999      0.886      0.844       0.89      0.654

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     53/300      1.58G     0.8007     0.7543      1.108         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.16it/s]


                   all       1720       3999      0.888      0.848      0.891       0.66

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     54/300      1.58G     0.8038     0.7501      1.112         18        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:20<00:00, 10.31it/s]


                   all       1720       3999      0.889      0.848      0.892      0.648

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     55/300      1.66G     0.7949      0.744      1.107         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.18it/s]


                   all       1720       3999      0.899      0.839      0.891      0.639

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     56/300      1.57G     0.7934     0.7356      1.104         19        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.23it/s]


                   all       1720       3999      0.909      0.837      0.893      0.643

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     57/300      1.66G     0.7918     0.7375      1.102         19        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.23it/s]


                   all       1720       3999      0.898      0.841      0.894      0.647

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     58/300      1.58G       0.79     0.7349      1.105         19        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.21it/s]


                   all       1720       3999      0.902       0.84      0.893      0.654

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     59/300      1.57G      0.787     0.7232      1.097         19        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.15it/s]


                   all       1720       3999      0.901       0.84      0.893      0.649

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     60/300      1.57G     0.7984     0.7289      1.107         19        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.23it/s]


                   all       1720       3999      0.895      0.849      0.894       0.66

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     61/300      1.66G     0.7841     0.7198      1.092         19        384: 100%|██████████| 1011/1011 [02:48<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.18it/s]


                   all       1720       3999      0.892      0.848      0.892       0.66

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     62/300      1.66G     0.7704     0.7124      1.088         19        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:20<00:00, 10.27it/s]


                   all       1720       3999      0.902      0.844      0.893      0.659

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


     63/300      1.58G     0.7757     0.7108      1.093         19        384: 100%|██████████| 1011/1011 [02:47<00:00,  6.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:21<00:00, 10.19it/s]


                   all       1720       3999      0.902      0.842      0.892      0.648
EarlyStopping: Training stopped early as no improvement observed in last 10 epochs. Best results observed at epoch 53, best model saved as best.pt.
To update EarlyStopping(patience=10) pass a new patience value, i.e. `patience=300` or use `patience=0` to disable EarlyStopping.

63 epochs completed in 3.335 hours.
Optimizer stripped from results/yolov11n_v1_ep300_img640_bs4_lr1e-05_lrf0.01_Adam_augmented_20250528_2232/weights/last.pt, 5.4MB
Optimizer stripped from results/yolov11n_v1_ep300_img640_bs4_lr1e-05_lrf0.01_Adam_augmented_20250528_2232/weights/best.pt, 5.4MB

Validating results/yolov11n_v1_ep300_img640_bs4_lr1e-05_lrf0.01_Adam_augmented_20250528_2232/weights/best.pt...
Ultralytics 8.3.141 🚀 Python-3.10.12 torch-2.7.0+cu126 CUDA:0 (NVIDIA GeForce MX550, 1864MiB)
YOLO11n summary (fused): 100 layers, 2,582,542 parameters, 0 gradients, 6.3 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 215/215 [00:18<00:00, 11.42it/s]


                   all       1720       3999      0.896      0.849      0.894       0.66
                person       1355       2841      0.808      0.718      0.796      0.478
                teleco        644       1158      0.984      0.979      0.992      0.842
Speed: 0.4ms preprocess, 8.6ms inference, 0.0ms loss, 0.6ms postprocess per image
Results saved to results/yolov11n_v1_ep300_img640_bs4_lr1e-05_lrf0.01_Adam_augmented_20250528_2232
✅ Trained checkpoint saved to: results/yolov11n_v1_ep300_img640_bs4_lr1e-05_lrf0.01_Adam_augmented_20250528_2232/weights/best.pt


# 🧠 YOLOv8 Training Metrics Cheatsheet

## 🔢 Loss Curves

| Metric            | Meaning                                                | Goal / Good Trend                |
|-------------------|--------------------------------------------------------|----------------------------------|
| `train/box_loss`  | Bounding box regression loss (training)                | ↓ Steady decline                 |
| `train/cls_loss`  | Classification loss (training)                         | ↓ Should fall and stabilize     |
| `train/dfl_loss`  | Distribution Focal Loss (for bounding box quality)     | ↓ Smooth drop                   |
| `val/box_loss`    | Box loss on validation set                             | ↓ Best indicator of generalization |
| `val/cls_loss`    | Classification loss on validation                      | ↓ Falling → Good model fit      |
| `val/dfl_loss`    | DFL loss on validation                                 | ↓ Matches train trend = no overfit |

---

## 📈 Detection Metrics

| Metric                | Description                                                 | Goal / Interpretation           |
|------------------------|-------------------------------------------------------------|----------------------------------|
| `metrics/precision(B)` | % of correct detections out of all predictions              | ↑ High = few false positives     |
| `metrics/recall(B)`    | % of actual objects detected                                | ↑ High = few missed detections   |
| `metrics/mAP50(B)`     | Accuracy of boxes at 50% IoU threshold                     | ↑ Main object detection metric   |
| `metrics/mAP50-95(B)`  | Average mAP over IoU 0.5 to 0.95 (stricter)                | ↑ Robust performance measure     |

---

## 🧪 Ideal Curve Patterns

- **Losses**: should steadily decrease.
- **Validation loss** rising while training loss decreases → possible **overfitting**.
- **Precision & Recall**: should increase and plateau near 1.0.
- **mAP**: aim for **>0.90 mAP50** and **>0.80 mAP50-95** for high-quality results.

---

## ⚠️ Tips

- 🧊 Use `half=True` for mixed precision training on supported GPUs.
- 🧠 Watch `val/box_loss` and `mAP50-95` to catch overfitting early.
- 🧪 If metrics plateau early, try increasing `epochs`, `imgsz`, or improving labels.
- 💾 Check `runs/train/.../results.png` for curve visualization.

---
